software/common/libs/framing/frame_detector.tpp

```diff
@@ -0,0 +1,165 @@
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+////////////////////////////////////////////////////////////////////////////////
+/// WARNING: Do not include this file as header, nor compile it as a source! ///
+///                                                                          ///
+/// This is a template implementation file.                                  ///
+/// It is included by the corresponding header.                              ///
+////////////////////////////////////////////////////////////////////////////////
+
+template<size_t S>
+FrameDetector<S>::FrameDetector(RxBuffer* buffer) : rx_buffer_(buffer)
+{}
+
+template<size_t S>
+bool FrameDetector<S>::begin()
+{
+  state_ = State::Lost;
+  frame_available_ = false;
+  return rx_buffer_->begin(this);
+}
+
+template<size_t S>
+void FrameDetector<S>::on_char_match_with_data()
+{
+  switch (state_)
+  {
+    case State::Lost:
+      // We have received something more before this marker.
+      // We assume this is the frame end marker, so we wait for start.
+      state_ = State::WaitForStartMarker;
+      restart_rx();
+      return;
+    case State::WaitForStartMarker:
+      // Received some junk while waiting for start marker but should have been just silence.
+      state_ = State::Lost;
+      restart_rx();
+      return;
+    case State::ReceivingFrame:
+      // Yes, we got data, thus we got the frame we can pass further.
+      process_received_data();
+      state_ = State::WaitForStartMarker;
+      restart_rx();
+      return;
+  }
+  // Switch above covers all cases.
+  __builtin_unreachable();
+}
+
+template<size_t S>
+void FrameDetector<S>::on_char_match_with_markers_only()
+{
+  switch (state_)
+  {
+    case State::Lost:
+      // Received just this marker.
+      // We were lucky to get to Lost state in the inter-frame silence,
+      // so we assume this marker is the start of frame.
+      state_ = State::ReceivingFrame;
+      return;
+    case State::WaitForStartMarker:
+      // Received the marker we are waiting for.
+      state_ = State::ReceivingFrame;
+      return;
+    case State::ReceivingFrame:
+      // Received repeated marker char.
+      // We have received a 0-length frame. Ignore it and continue receiving frame bytes.
+      restart_rx();
+      return;
+  }
+  // Switch above covers all cases.
+  __builtin_unreachable();
+}
+
+template<size_t S>
+void FrameDetector<S>::on_character_match()
+{
+  marker_count_++;
+
+  if (data_received()) {
+    on_char_match_with_data();
+  } else if (only_markers_received()) {
+    on_char_match_with_markers_only();
+  } else {
+    // Should never end up here
+    // DMA is not working?
+    // TODO alert, safe reset
```

Comment:
Same comment apply as for PR 1223 about using Doxygen markup and use \TODO (or \todo) to automatically compile a list of todo's